File: Jan2021/03Jan/BestTimeToBuyAndSellStock.cpp
/*
Problem name: Best Time to Buy and Sell Stock
Problem link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
*/


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        if(n<2)
            return 0;
        int maxprofit=0, rightmax=prices[n-1];
        
        for(int i=n-2 ;i>=0 ;--i){
            maxprofit=max(maxprofit, rightmax-prices[i]);
            rightmax=max(rightmax, prices[i]);
        }
        return maxprofit;
    }
};

File: Jan2021/03Jan/BestTimeToBuyAndSellStock.java
/*
Problem Name: 121. Best Time to Buy and Sell Stock
Problem Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
*/
class Solution {
    public int maxProfit(int[] prices) {
        int maxProf = 0, n = prices.length;
        for(int i=1;i<n;i++){
            maxProf = Math.max(maxProf, prices[i] - prices[i-1]);
            prices[i] = Math.min(prices[i],prices[i-1]);
        }
        return maxProf;
    }
}
